File: src/main/java/com/barbaros/expense/service/AuthService.java
package com.barbaros.expense.service;

import com.barbaros.expense.dto.AuthDTO;
import com.barbaros.expense.model.User;
import com.barbaros.expense.repository.UserRepository;
import com.barbaros.expense.security.JwtUtil;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class AuthService {

    private final UserRepository userRepository;
    private final JwtUtil jwtUtil;
    private final BCryptPasswordEncoder passwordEncoder;

    public AuthService(UserRepository userRepository, JwtUtil jwtUtil) {
        this.userRepository = userRepository;
        this.jwtUtil = jwtUtil;
        this.passwordEncoder = new BCryptPasswordEncoder();
    }

    public String register(AuthDTO request) {
        Optional<User> existingUser = Optional.ofNullable(userRepository.findByUsername(request.getUsername()));
        if (existingUser.isPresent()) {
            throw new IllegalArgumentException("Email already in use");
        }

        User user = new User();
        user.setUsername(request.getUsername());
        user.setPassword(passwordEncoder.encode(request.getPassword()));

        userRepository.save(user);
        return "User registered successfully!";
    }

    public String login(AuthDTO request) {
        User user = userRepository.findByUsername(request.getUsername());

        if (user == null) {
            throw new IllegalArgumentException("Invalid credentials");
        }

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new IllegalArgumentException("Invalid credentials");
        }

        return jwtUtil.generateToken(user.getUsername());
    }
}

File: src/main/kotlin/com/barbaros/expense/service/AuthService.kt
package com.barbaros.expense.service

import com.barbaros.expense.dto.LoginDTO
import com.barbaros.expense.dto.RegisterDTO
import com.barbaros.expense.model.User
import com.barbaros.expense.repository.UserRepository
import com.barbaros.expense.security.JwtUtil
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.stereotype.Service
import java.lang.IllegalArgumentException

@Service
class AuthService(
    private val userRepository: UserRepository,
    private val jwtUtil: JwtUtil
) {
    private val passwordEncoder = BCryptPasswordEncoder()

    fun register(request: RegisterDTO): String {
        if (userRepository.findByEmail(request.email).isPresent) {
            throw IllegalArgumentException("Email already in use")
        }

        val user = User(
            email =  request.email,
            password = passwordEncoder.encode(request.password)
        )

        userRepository.save(user)

        return "User registered successfully!"
    }

    fun login(request: LoginDTO): String {
        val user = userRepository.findByEmail(request.email)
            .orElseThrow { IllegalArgumentException("User not found") }

        if (!passwordEncoder.matches(request.password, user.password)) {
            throw IllegalArgumentException("Invalid credentials")
        }

        return jwtUtil.generateToken(user.email)
    }
}
